pkg/controller/actions/observability/action_monitoring.go

```diff
@@ -0,0 +1,97 @@
+package observability
+
+import (
+	"context"
+	"fmt"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/kustomize/api/builtins"
+	kTypes "sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/resid"
+
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/actions"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/types"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/manifests/kustomize"
+)
+
+const ObservabilityFieldOwner = "observability"
+
+type Action struct {
+	labels      map[string]string
+	namespaceFn actions.StringGetter
+}
+
+type ActionOpts func(*Action)
+
+func NewAction(opts ...ActionOpts) actions.Fn {
+	action := Action{
+		labels:      map[string]string{},
+		namespaceFn: actions.ApplicationNamespace,
+	}
+
+	for _, opt := range opts {
+		opt(&action)
+	}
+
+	return action.run
+}
+
+func WithLabel(name string, value string) ActionOpts {
+	return func(action *Action) {
+		if action.labels == nil {
+			action.labels = map[string]string{}
+		}
+		action.labels[name] = value
+	}
+}
+
+func (a *Action) run(ctx context.Context, rr *types.ReconciliationRequest) error {
+	namespace, err := a.namespaceFn(ctx, rr)
+	if err != nil {
+		return fmt.Errorf("failed to get namespace: %w", err)
+	}
+	// Using CreateLabelsTransformer to set labels avoiding pod selectors
+	labelPlugin := CreateLabelsTransformer(a.labels)
+	engine := kustomize.NewEngine(
```

Comment:
I think you can probably just iterate through the` rr.Resources` slice and patch the deployments instead of re-render the manifests, then let the deploy action to actually patch the object if needed:

```go
rr.ForEachResource(func(obj *unstructured.Unstructured) (bool, error) {
    // return if not a Deployment
    // add label to the obj
    // add label to the spec/template/metadata/labels
    return false, nil    
})
```